storage/client_range_test.go

```diff
@@ -150,3 +154,156 @@ func TestRejectFutureCommand(t *testing.T) {
 		t.Errorf("expected 15, got %v", v)
 	}
 }
+
+// Test a case where a put operation of an older timestamp comes after
+// a put operation of a newer timestamp.
+//
+// 1) Writer executes a put operation with time T in a txn.
+// 2) Before the txn is committed, Reader sends a get
+//    operation with time T+100. This triggers the txn restart.
+// 3) Reader sends another get operation with time T+200. The
+//    write intent is resolved (and the txn timestamp is pushed to
+//    T+200), but the actual get operation has not yet been executed (and
+//    hence the timestamp cache has not been updated).
+// 4) Writer restarts the txn and executes the put operation
+//    again. The timestamp of the operation is T+100, and it will be
+//    ignored.
+func TestOutOfOrderPut(t *testing.T) {
```

Comment:
rename to `TestTxnPutOutOfOrder`? There's `TestMVCCPutOutOfOrder`, which from the looks of it already tests what you have here at the MVCC level.